File: app/interceptor/crud.py
"""Auto-CRUD engine over Redis-backed JSON arrays (arch §4.3, §5.5; AC-11/12/12a/12b).

When ``auto_crud`` is enabled and **no rule matches**, the interceptor turns the
endpoint into a zero-config REST DB backend over a Redis list per collection
(``crud:<token>:<collection>``). The frozen REST lifecycle (§5.5 / AC-12):

    POST   /<collection>        -> 201  create (server-generated uuid4 ``id``)
    GET    /<collection>        -> 200  list (the whole array)
    GET    /<collection>/<id>   -> 200 one | 404
    PUT    /<collection>/<id>   -> 200 replace (keeps ``id``) | 404
    PATCH  /<collection>/<id>   -> 200 shallow-merge | 404
    DELETE /<collection>/<id>   -> 204 | 404
    (HEAD mirrors GET; OPTIONS never reaches here — handled by the CORS preflight.)

Validation & bounds (AC-12a / AC-12b / SEC-AC-21/24):
  * ``<collection>`` and ``<id>`` must match ``^[A-Za-z0-9_-]{1,64}$`` (the same
    safe charset as state keys) so a crafted segment cannot inject a Redis
    separator and escape the endpoint namespace; a bad segment is treated as a
    non-CRUD path (``matches`` returns False → falls through to MITM/default).
  * A write body must be a JSON **object**; invalid / empty / non-object / oversize
    (> ``CRUD_MAX_ITEM_BYTES``) bodies are rejected with **400** (never 500).
  * The collection is bounded by ``CRUD_MAX_ITEMS`` (enforced in the Redis facade)
    and per-item bytes by ``CRUD_MAX_ITEM_BYTES``; exceeding either is **400**.

Redis-down (§5.11 / AC-49): every store op raises :class:`RedisUnavailable`, which
the engine maps to **503** (we never fabricate or silently lose data). The id is a
``uuid4`` so concurrent ``POST`` s never collide (AC-12a).

The engine resolution order keeps **rules > CRUD** (a rule matching ``/<collection>``
wins and this module is never consulted for that request), and CRUD is rate-limited
on writes by the engine's ``eff_rate`` token bucket (AC-S19).
"""

from __future__ import annotations

import json
import logging
import re
import uuid
from typing import List, Optional, Tuple

from fastapi.responses import JSONResponse, Response

import config
from app.redis_state import RedisUnavailable, redis_state
from app.utils.helpers import is_safe_key

logger = logging.getLogger("hookbox.crud")

# A CRUD path is ``/<collection>`` or ``/<collection>/<id>``. Anything deeper (or
# with empty/oversize/unsafe segments) is NOT CRUD and falls through to the next
# resolution stage. We validate the charset of each segment with ``is_safe_key``.
_WRITE_METHODS = {"POST", "PUT", "PATCH", "DELETE"}

# crud_cas result sentinels (distinguish not-found / too-large from a returned item).
_NOT_FOUND = object()
_TOO_LARGE = object()
_OK = object()


def _parse_path(mock_path: str) -> Optional[Tuple[str, Optional[str]]]:
    """Parse ``/<collection>[/<id>]`` → ``(collection, id|None)``.

    Returns ``None`` when the path is not a valid 1- or 2-segment CRUD path whose
    segments pass the safe charset (so the caller treats it as a non-CRUD path).
    """
    if not mock_path or not mock_path.startswith("/"):
        return None
    # Drop a single trailing slash so ``/books/`` == ``/books``.
    trimmed = mock_path[:-1] if mock_path.endswith("/") and len(mock_path) > 1 else mock_path
    segs = [s for s in trimmed.split("/") if s != ""]
    if len(segs) == 1:
        coll = segs[0]
        if not is_safe_key(coll):
            return None
        return coll, None
    if len(segs) == 2:
        coll, ident = segs
        if not is_safe_key(coll) or not is_safe_key(ident):
            return None
        return coll, ident
    return None


def matches(mock_path: str) -> bool:
    """True iff ``mock_path`` looks like a CRUD collection/item path (AC-12).

    Used by the engine to decide whether to consult Auto-CRUD before MITM/default.
    A 3+ segment path or an unsafe segment is **not** CRUD (returns False).
    """
    return _parse_path(mock_path) is not None


def _bad_request(detail: str) -> JSONResponse:
    return JSONResponse(status_code=400, content={"error": "bad_request", "detail": detail})


def _not_found() -> JSONResponse:
    return JSONResponse(status_code=404, content={"error": "not_found",
                        "detail": "No such item."})


def _parse_object_body(body_text: str) -> dict:
    """Parse a write body into a JSON object or raise ``ValueError`` (→ 400).

    Enforces the per-item byte cap (AC-12b) before parsing.
    """
    if body_text is None or body_text.strip() == "":
        raise ValueError("empty body")
    if len(body_text.encode("utf-8")) > config.CRUD_MAX_ITEM_BYTES:
        raise ValueError("item too large")
    try:
        obj = json.loads(body_text)
    except (ValueError, TypeError) as exc:
        raise ValueError("body is not valid JSON") from exc
    if not isinstance(obj, dict):
        raise ValueError("body must be a JSON object")
    return obj


async def handle(token: str, method: str, mock_path: str, body_text: str) -> Response:
    """Execute one Auto-CRUD operation (§5.5 / AC-12).

    Raises :class:`RedisUnavailable` on Redis loss so the engine returns 503
    (AC-49). All bad-input paths return a 400/404 JSONResponse (never 500).
    """
    parsed = _parse_path(mock_path)
    if parsed is None:
        # Defensive: the engine only calls us when ``matches`` is True, but never
        # 500 if it changes — treat as no-match → 404.
        return _not_found()
    collection, ident = parsed
    method = method.upper()

    # --- collection-level (no id) --------------------------------------------
    if ident is None:
        if method in ("GET", "HEAD"):
            items = await redis_state.crud_list(token, collection)
            return JSONResponse(status_code=200, content=items)
        if method == "POST":
            try:
                obj = _parse_object_body(body_text)
            except ValueError as exc:
                return _bad_request(str(exc))
            # Server-assigned id wins (collision-safe uuid4, AC-12a). A client id
            # is ignored/overwritten so two POSTs never collide.
            obj["id"] = uuid.uuid4().hex
            try:
                await redis_state.crud_append(token, collection, obj)
            except RedisUnavailable:
                raise
            except ValueError:
                # CRUD_MAX_ITEMS reached (collection_full) — bounded, not 500.
                return _bad_request("collection is full")
            return JSONResponse(status_code=201, content=obj)
        # PUT/PATCH/DELETE on a collection (no id) is not a CRUD item op.
        return _bad_request("an item id is required for this method")

    # --- item-level (collection + id) ----------------------------------------
    # GET/HEAD are read-only; the mutating verbs use an ATOMIC read-modify-write
    # (redis_state.crud_cas, WATCH/MULTI + retry) so two concurrent writers to the
    # same collection can't lose an update (hookbox-65m). Input is parsed/validated
    # (→ 400) BEFORE the transaction; not-found / too-large come back via sentinels.
    if method in ("GET", "HEAD"):
        items = await redis_state.crud_list(token, collection)
        idx = _find_index(items, ident)
        if idx is None:
            return _not_found()
        return JSONResponse(status_code=200, content=items[idx])

    if method == "PUT":
        try:
            obj = _parse_object_body(body_text)
        except ValueError as exc:
            return _bad_request(str(exc))
        obj["id"] = ident  # id is immutable on replace

        def _put(items):
            i = _find_index(items, ident)
            if i is None:
                return _NOT_FOUND, None
            items[i] = dict(obj)
            return items[i], items

        result = await redis_state.crud_cas(token, collection, _put)
        if result is _NOT_FOUND:
            return _not_found()
        return JSONResponse(status_code=200, content=result)

    if method == "PATCH":
        try:
            patch = _parse_object_body(body_text)
        except ValueError as exc:
            return _bad_request(str(exc))

        def _patch(items):
            i = _find_index(items, ident)
            if i is None:
                return _NOT_FOUND, None
            merged = dict(items[i])
            merged.update(patch)
            merged["id"] = ident  # id is immutable on merge
            if len(json.dumps(merged).encode("utf-8")) > config.CRUD_MAX_ITEM_BYTES:
                return _TOO_LARGE, None
            items[i] = merged
            return merged, items

        result = await redis_state.crud_cas(token, collection, _patch)
        if result is _NOT_FOUND:
            return _not_found()
        if result is _TOO_LARGE:
            return _bad_request("item too large")
        return JSONResponse(status_code=200, content=result)

    if method == "DELETE":
        def _delete(items):
            i = _find_index(items, ident)
            if i is None:
                return _NOT_FOUND, None
            del items[i]
            return _OK, items

        result = await redis_state.crud_cas(token, collection, _delete)
        if result is _NOT_FOUND:
            return _not_found()
        return Response(status_code=204)

    return _bad_request("unsupported method for an item")


def _find_index(items: List[dict], ident: str) -> Optional[int]:
    """Return the list index whose element ``id`` == ``ident``, else None."""
    for i, it in enumerate(items):
        if isinstance(it, dict) and str(it.get("id")) == ident:
            return i
    return None

File: backend/src/interceptor/crud.rs
//! Auto-CRUD path routing — PORT of `app/interceptor/crud.py` (§5.5, AC-24..27).
//!
//! When `auto_crud` is on and no rule matched, the interceptor turns the endpoint
//! into a zero-config REST DB over `crud_collections`. The frozen lifecycle:
//!   POST /<coll>        -> 201 (server uuid id)
//!   GET  /<coll>        -> 200 array
//!   GET  /<coll>/<id>   -> 200 | 404
//!   PUT  /<coll>/<id>   -> 200 | 404
//!   PATCH/<coll>/<id>   -> 200 | 404
//!   DELETE /<coll>/<id> -> 204 | 404   (HEAD mirrors GET)
//!
//! Segments must match `^[A-Za-z0-9_-]{1,64}$`; a 3+-segment or unsafe path is NOT
//! CRUD (`matches` is false) → the engine falls through to tunnel/MITM/default.
//! A write body must be a JSON object (else 400); per-item byte cap → 400; the
//! collection item cap → 400. A SQLite fault surfaces as 503.

use serde_json::Value;

use crate::config::Config;
use crate::crud_store::{self, Mutation};
use crate::helpers::is_safe_key;
use crate::state::AppState;

/// A CRUD outcome the engine renders into an HTTP response (served_by="crud").
pub struct CrudResponse {
    pub status: u16,
    /// `None` for a 204 (no body).
    pub body: Option<Value>,
}

impl CrudResponse {
    fn json(status: u16, body: Value) -> Self {
        CrudResponse { status, body: Some(body) }
    }
    fn no_content() -> Self {
        CrudResponse { status: 204, body: None }
    }
    fn bad_request(detail: &str) -> Self {
        CrudResponse::json(400, serde_json::json!({"error": "bad_request", "detail": detail}))
    }
    fn not_found() -> Self {
        CrudResponse::json(404, serde_json::json!({"error": "not_found", "detail": "No such item."}))
    }
}

/// Parse `/<coll>[/<id>]` into `(coll, id?)`; `None` when not a valid 1- or
/// 2-segment safe CRUD path.
pub fn parse_path(mock_path: &str) -> Option<(String, Option<String>)> {
    if mock_path.is_empty() || !mock_path.starts_with('/') {
        return None;
    }
    let trimmed = if mock_path.len() > 1 && mock_path.ends_with('/') {
        &mock_path[..mock_path.len() - 1]
    } else {
        mock_path
    };
    let segs: Vec<&str> = trimmed.split('/').filter(|s| !s.is_empty()).collect();
    match segs.len() {
        1 if is_safe_key(segs[0]) => Some((segs[0].to_string(), None)),
        2 if is_safe_key(segs[0]) && is_safe_key(segs[1]) => {
            Some((segs[0].to_string(), Some(segs[1].to_string())))
        }
        _ => None,
    }
}

/// True iff `mock_path` looks like a CRUD collection/item path (AC-24).
pub fn matches(mock_path: &str) -> bool {
    parse_path(mock_path).is_some()
}

fn parse_object_body(body_text: &str, cfg: &Config) -> Result<serde_json::Map<String, Value>, String> {
    if body_text.trim().is_empty() {
        return Err("empty body".into());
    }
    if body_text.len() > cfg.crud_max_item_bytes {
        return Err("item too large".into());
    }
    let v: Value = serde_json::from_str(body_text).map_err(|_| "body is not valid JSON".to_string())?;
    match v {
        Value::Object(m) => Ok(m),
        _ => Err("body must be a JSON object".into()),
    }
}

fn find<'a>(items: &'a [Value], id: &str) -> Option<&'a Value> {
    items.iter().find(|it| it.get("id").and_then(|v| v.as_str()) == Some(id))
}

/// Execute one Auto-CRUD operation. Returns a `CrudResponse`, or `Err(())` on a
/// SQLite fault (the engine maps to 503 store_unavailable).
pub async fn handle(
    state: &AppState,
    token: &str,
    method: &str,
    mock_path: &str,
    body_text: &str,
) -> Result<CrudResponse, ()> {
    let (coll, ident) = match parse_path(mock_path) {
        Some(p) => p,
        None => return Ok(CrudResponse::not_found()),
    };
    let method = method.to_ascii_uppercase();
    let cfg = &state.cfg;
    let ttl = cfg.crud_ttl_seconds;

    // Collection-level (no id).
    if ident.is_none() {
        match method.as_str() {
            "GET" | "HEAD" => {
                let items = crud_store::list_items(&state.pool, token, &coll).await.map_err(|_| ())?;
                return Ok(CrudResponse::json(200, Value::Array(items)));
            }
            "POST" => {
                let obj = match parse_object_body(body_text, cfg) {
                    Ok(o) => o,
                    Err(e) => return Ok(CrudResponse::bad_request(&e)),
                };
                return match crud_store::append_item(&state.pool, token, &coll, obj, cfg.crud_max_items, ttl).await {
                    Ok(Mutation::Ok(_, item)) => Ok(CrudResponse::json(201, item)),
                    Ok(Mutation::TooLarge) => Ok(CrudResponse::bad_request("collection is full")),
                    Ok(Mutation::NotFound) => Ok(CrudResponse::not_found()),
                    Err(_) => Err(()),
                };
            }
            _ => return Ok(CrudResponse::bad_request("an item id is required for this method")),
        }
    }

    // Item-level (collection + id).
    let id = ident.unwrap();
    match method.as_str() {
        "GET" | "HEAD" => {
            let items = crud_store::list_items(&state.pool, token, &coll).await.map_err(|_| ())?;
            match find(&items, &id) {
                Some(item) => Ok(CrudResponse::json(200, item.clone())),
                None => Ok(CrudResponse::not_found()),
            }
        }
        "PUT" => {
            let obj = match parse_object_body(body_text, cfg) {
                Ok(o) => o,
                Err(e) => return Ok(CrudResponse::bad_request(&e)),
            };
            match crud_store::replace_item(&state.pool, token, &coll, &id, obj, ttl).await {
                Ok(Mutation::Ok(_, item)) => Ok(CrudResponse::json(200, item)),
                Ok(Mutation::NotFound) => Ok(CrudResponse::not_found()),
                Ok(Mutation::TooLarge) => Ok(CrudResponse::bad_request("item too large")),
                Err(_) => Err(()),
            }
        }
        "PATCH" => {
            let patch = match parse_object_body(body_text, cfg) {
                Ok(o) => o,
                Err(e) => return Ok(CrudResponse::bad_request(&e)),
            };
            match crud_store::merge_item(&state.pool, token, &coll, &id, patch, cfg.crud_max_item_bytes, ttl).await {
                Ok(Mutation::Ok(_, item)) => Ok(CrudResponse::json(200, item)),
                Ok(Mutation::NotFound) => Ok(CrudResponse::not_found()),
                Ok(Mutation::TooLarge) => Ok(CrudResponse::bad_request("item too large")),
                Err(_) => Err(()),
            }
        }
        "DELETE" => match crud_store::delete_item(&state.pool, token, &coll, &id, ttl).await {
            Ok(Mutation::Ok(..)) => Ok(CrudResponse::no_content()),
            Ok(Mutation::NotFound) => Ok(CrudResponse::not_found()),
            Ok(Mutation::TooLarge) => Ok(CrudResponse::bad_request("item too large")),
            Err(_) => Err(()),
        },
        _ => Ok(CrudResponse::bad_request("unsupported method for an item")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_path_rules() {
        assert_eq!(parse_path("/books"), Some(("books".into(), None)));
        assert_eq!(parse_path("/books/"), Some(("books".into(), None)));
        assert_eq!(parse_path("/books/abc"), Some(("books".into(), Some("abc".into()))));
        // 3 segments -> not CRUD
        assert_eq!(parse_path("/books/abc/x"), None);
        // unsafe segment -> not CRUD
        assert_eq!(parse_path("/bad name"), None);
        assert_eq!(parse_path("/books/bad id"), None);
        assert!(matches("/books"));
        assert!(!matches("/a/b/c"));
    }
}
